# Module vs Library

| Feature             | Module                                   | Library                                       |
|--------------------|-----------------------------------------|-----------------------------------------------|
| Definition          | A single Python file (.py) containing functions, classes, or variables | A collection of modules grouped together to provide related functionality |
| Example             | `math.py`, `datetime.py`, `json.py`     | Python Standard Library (contains math, datetime, json, etc.) |
| Purpose             | Organize code into a single file for reuse | Organize related modules to provide more complete functionality |
| Size/Scope          | Small (one file)                         | Larger (many modules in a package or library) |
| Usage               | Imported using `import module_name`      | Imported using `import module_name` (modules inside the library) |
| Custom or Standard  | Can be standard or user-created          | Usually standard library or external library (like NumPy) |

**Note:**  
- In Python, when you say “library,” it often refers to a **collection of modules**, not necessarily a single file.  
- Modules are the building blocks of libraries.

# Python Daytime

## Python Dates

A date in Python is not a data type of its own, but we can import a module named `datetime` to work with dates as data objects

Eg: Import the datetime module and display the current date:

In [1]:
import datetime

x = datetime.datetime.now()
print(x)

2025-09-24 15:22:21.349006


## Date O/P

When we execute the code from the example above the result will be:

`2025-09-24 15:22:21.349006`

The date contains year, month, day, hour, minute, second, and microsecond.

The datetime module has many methods to return information about the date object.


Eg: return the year and name of weekday:

In [ ]:
import datetime

x = datetime.datetime.now()

print(x.year)

# strftime() formats date/time into a readable string
# %A gives the full weekday name (e.g., Monday, Tuesday)
print(x.strftime("%A"))

2025
Wednesday



## ===============================
## DATETIME Cheat Sheet
## ===============================

### 1. strftime() -> datetime object to formatted string
####    Use to display a datetime in a readable format
####    Example: x.strftime("%A, %d-%B-%Y %H:%M:%S")

## Common format codes:
```python
%a  -> Abbreviated weekday name      (Mon, Tue)
%A  -> Full weekday name             (Monday, Tuesday)
%w  -> Weekday as number             (0=Sunday, 6=Saturday)

%d  -> Day of month (01-31)
%b  -> Abbreviated month name        (Jan, Feb)
%B  -> Full month name               (January, February)
%m  -> Month as number (01-12)

%y  -> Year without century          (23 for 2023)
%Y  -> Year with century             (2023)

%H  -> Hour (24-hour, 00-23)
%I  -> Hour (12-hour, 01-12)
%p  -> AM/PM

%M  -> Minute (00-59)
%S  -> Second (00-59)
%f  -> Microsecond (000000-999999)
%j  -> Day of the year (001-366)
%U  -> Week number (Sunday first)
%W  -> Week number (Monday first)
```

### 2. strptime() -> string to datetime object
###    Use to parse a string representing a date/time into a datetime object
###    Example:
###    date_string = "24-09-2025 13:30"
###    date_object = datetime.datetime.strptime(date_string, "%d-%m-%Y %H:%M")
###    print(date_object)  # 2025-09-24 13:30:00

### Key difference:
#### - strftime() = datetime -> string
#### - strptime() = string -> datetime
#### - Both use the same format codes above

In [3]:
import datetime

# strptime() = "string parse time"
# It converts a string representing a date/time into a datetime object
# You must provide the format of the string using the same codes as strftime

date_string = "24-09-2025 13:30"
date_object = datetime.datetime.strptime(date_string, "%d-%m-%Y %H:%M")

print(date_object)  # 2025-09-24 13:30:00

2025-09-24 13:30:00


## Creating Date Objects

To create a date, we can use the `datetime()` class(constructor) of the `datetime` module.

The `datetime()` class requires three parameters to create a date: year, month, day.

In [4]:
import datetime
x = datetime.datetime(2024, 2, 19)

print(x)

2024-02-19 00:00:00


The `datetime()` class also takes parameters for time and timezone (hour, minute, second, microsecond, tzone), but they are optional, and has a default value of `0`, (`None` for timezone).

## The strftime() Method

The `datetime` object has a method for formatting date objects into readable strings.

The method is called `strftime()`, and takes one parameter, `format`, to specify the format of the returned string:

Eg: Display the name of the month:

In [5]:
import datetime

x = datetime.datetime(2025, 2, 28)

print(x.strftime("%B"))

February
